pkg/autodiscovery/listeners/kube_services_test.go

```diff
@@ -251,9 +251,382 @@ func TestServicesDiffer(t *testing.T) {
 		},
+		"Add endpoints annotations": {
+			first: &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					ResourceVersion: "123",
+					Annotations: map[string]string{
+						"ad.datadoghq.com/service.check_names":  "[\"http_check\"]",
+						"ad.datadoghq.com/service.init_configs": "[{}]",
+						"ad.datadoghq.com/service.instances":    "[{\"name\": \"My service\", \"url\": \"http://%%host%%\", \"timeout\": 1}]",
+					},
+				},
+				Spec: v1.ServiceSpec{
+					ClusterIP: "10.0.0.1",
+					Ports: []v1.ServicePort{
+						{Name: "test2", Port: 126},
+					},
+				},
+			},
+			second: &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					ResourceVersion: "124",
+					Annotations: map[string]string{
+						"ad.datadoghq.com/service.check_names":    "[\"http_check\"]",
+						"ad.datadoghq.com/service.init_configs":   "[{}]",
+						"ad.datadoghq.com/service.instances":      "[{\"name\" \"My service\", \"url\": \"http://%%host%%\", \"timeout\": 1}]",
+						"ad.datadoghq.com/endpoints.check_names":  "[\"etcd\"]",
+						"ad.datadoghq.com/endpoints.init_configs": "[{}]",
+						"ad.datadoghq.com/endpoints.instances":    "[{\"use_preview\": \"true\", \"prometheus_url\": \"http://%%host%%:2379/metrics\"}]",
+					},
+				},
+				Spec: v1.ServiceSpec{
+					ClusterIP: "10.0.0.1",
+					Ports: []v1.ServicePort{
+						{Name: "test2", Port: 126},
+					},
+				},
+			},
+			result: true,
+		},
+		"Remove endpoints annotations": {
+			first: &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					ResourceVersion: "123",
+					Annotations: map[string]string{
+						"ad.datadoghq.com/service.check_names":    "[\"http_check\"]",
+						"ad.datadoghq.com/service.init_configs":   "[{}]",
+						"ad.datadoghq.com/service.instances":      "[{\"name\": \"My service\", \"url\": \"http://%%host%%\", \"timeout\": 1}]",
+						"ad.datadoghq.com/endpoints.check_names":  "[\"etcd\"]",
+						"ad.datadoghq.com/endpoints.init_configs": "[{}]",
+						"ad.datadoghq.com/endpoints.instances":    "[{\"use_preview\": \"true\", \"prometheus_url\": \"http://%%host%%:2379/metrics\"}]",
+					},
+				},
+				Spec: v1.ServiceSpec{
+					ClusterIP: "10.0.0.1",
+					Ports: []v1.ServicePort{
+						{Name: "test2", Port: 126},
+					},
+				},
+			},
+			second: &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					ResourceVersion: "124",
+					Annotations: map[string]string{
+						"ad.datadoghq.com/service.check_names":  "[\"http_check\"]",
+						"ad.datadoghq.com/service.init_configs": "[{}]",
+						"ad.datadoghq.com/service.instances":    "[{\"name\" \"My service\", \"url\": \"http://%%host%%\", \"timeout\": 1}]",
+					},
+				},
+				Spec: v1.ServiceSpec{
+					ClusterIP: "10.0.0.1",
+					Ports: []v1.ServicePort{
+						{Name: "test2", Port: 126},
+					},
+				},
+			},
+			result: true,
+		},
 	} {
 		t.Run(name, func(t *testing.T) {
 			assert.Equal(t, tc.result, servicesDiffer(tc.first, tc.second))
 		})
 	}
 }
+
+func TestProcessEndpoint(t *testing.T) {
```

Comment:
I'd love to try creating the following test, as a template for the lifecycle tests we'll need to write:

- init a KServiceListener struct manually
- pass an annotated v1.Service via `createSvc`
- pass the sibling v1.Endpoints with two addrs via `createEndpoints`
- test two new services arrive on the `newService` chan